### ISSUE TITLE
Deepar gets wired loss

### ISSUE THREAD

**Issue by creamiracle** [awslabs/gluon-ts#1880]
HI there, I got a problem when using Gluonts-deepar. The input data contains a year(2021)'s data, which all are 175.0, then I transferr them into listdataset and send to deepar estimator with predict_length = 365 and context_length = 180. I get a really wierd loss, which is 60+, and the results of next prediction seems like the model did not train at all.
Is there any idea about this situation?
Thanks.
------------------ CODE --------------------
def train_model(seq, epoch, lr):
    train_ds = ListDataset(seq,freq = "D")
    estimator = deepar.DeepAREstimator(
        prediction_length=365,
        context_length=180,
        num_parallel_samples=200,
        freq="D",
        trainer=Trainer(ctx="cpu", epochs=epoch, learning_rate=lr)
    )
    predictor = estimator.train(training_data=train_ds)
    return predictor

**Comment by lostella** [comment]
@creamiracle could you provide a self-contained example? For instance, a snippet that also calls your `train_model` function there with the right `seq`, `epoch`, and `lr` that result in a weird model. Also:
* use triple-backticks ``` to open/close code snippets, so they are more readable ([see here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code))
* what version of python/mxnet/gluonts do you have installed?

**Comment by creamiracle** [comment]
> @creamiracle could you provide a self-contained example? For instance, a snippet that also calls your `train_model` function there with the right `seq`, `epoch`, and `lr` that result in a weird model. Also:
> 
> * use triple-backticks ``` to open/close code snippets, so they are more readable ([see here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code))
> * what version of python/mxnet/gluonts do you have installed?

My bad, I will write it more clearly.
Here is a simple example.
```
# The function to generate sequence which contains a year (365 days) with same nums
def gen_train_seq():
    seq, tmp_dict = [], {} 
    tmp_dict['start'] = '2021-01-01'
    tmp_dict['target'] = [152]*365
    seq.append(tmp_dict)
    return seq

# The function for training model
def train_model(seq, prediction_length,context_length, epoch, lr):
    train_ds = ListDataset(seq,freq = "D")
    estimator = deepar.DeepAREstimator(
        prediction_length=prediction_length,
        context_length=context_length,
        num_parallel_samples=200,
        freq="D",
        trainer=Trainer(ctx="cpu", epochs=epoch, learning_rate=lr)
    )
    predictor = estimator.train(training_data=train_ds)
    return predictor

# The function for prediction
def _predict(predictor, seq):
    df = pd.DataFrame()
    test_ds = ListDataset(seq,freq = "D")
    pred = next(predictor.predict(test_ds)).mean
    ind = pd.date_range("2022-01-01","2022-12-31",freq="D")
    df['dt'] = ind.values
    df['pred'] = pred
    return df

# Then I use params like 
seq = gen_train_seq()
train_model(seq, 365, 180, 100, 1e-2)

# And I get the results as ( for saving time, I just post the first and the last round loss here and some of the predictions )
# loss log
# 100%|████████████████████████████████████████████████| 50/50 [00:40<00:00,  1.23it/s, epoch=1/100, avg_epoch_loss=57.6]
# 100%|██████████████████████████████████████████████| 50/50 [00:50<00:00,  1.02s/it, epoch=100/100, avg_epoch_loss=47.2]
# predictions
2022-01-01 | -8865.508789
2022-01-02 | -4789.898926
2022-01-03 | 20795.058594
```
And my version is
python = 3.7
mxnet = 1.7.0.post2
gluonts = 0.8.0

Thanks.

**Comment by lostella** [comment]
@creamiracle thanks! The issue seems to be with the length of your data vs `prediction_length`: both are 365, which means that the model training procedure will only be able to train the model to predict the *whole* data you give it, conditioned on *no* past data. If instead one had `prediction_length` smaller than your available data, the model will be able to look at many more `(past, future)` pairs to train on.

To verify this, you can lower `prediction_length` to 30 (I've also lowered the number of epochs in this case):

```python
predictor =train_model(seq, 30, 180, 5, 1e-2)
```

gives

```
100%|█████████████| 50/50 [00:13<00:00,  3.82it/s, epoch=1/5, avg_epoch_loss=4.38]
100%|█████████████| 50/50 [00:11<00:00,  4.53it/s, epoch=2/5, avg_epoch_loss=2.39]
100%|█████████████| 50/50 [00:11<00:00,  4.46it/s, epoch=3/5, avg_epoch_loss=1.95]
100%|█████████████| 50/50 [00:12<00:00,  3.97it/s, epoch=4/5, avg_epoch_loss=1.74]
100%|█████████████| 50/50 [00:12<00:00,  3.88it/s, epoch=5/5, avg_epoch_loss=1.58]
```

If you really need to train for `prediction_length=365`, I advise you to use longer training data (at least 2-3 years).

**Comment by creamiracle** [comment]
> @creamiracle thanks! The issue seems to be with the length of your data vs `prediction_length`: both are 365, which means that the model training procedure will only be able to train the model to predict the _whole_ data you give it, conditioned on _no_ past data. If instead one had `prediction_length` smaller than your available data, the model will be able to look at many more `(past, future)` pairs to train on.
> 
> To verify this, you can lower `prediction_length` to 30 (I've also lowered the number of epochs in this case):
> 
> ```python
> predictor = train_model(seq, 30, 180, 5, 1e-2)
> ```
> 
> gives
> 
> ```
> 100%|█████████████| 50/50 [00:13<00:00,  3.82it/s, epoch=1/5, avg_epoch_loss=4.38]
> 100%|█████████████| 50/50 [00:11<00:00,  4.53it/s, epoch=2/5, avg_epoch_loss=2.39]
> 100%|█████████████| 50/50 [00:11<00:00,  4.46it/s, epoch=3/5, avg_epoch_loss=1.95]
> 100%|█████████████| 50/50 [00:12<00:00,  3.97it/s, epoch=4/5, avg_epoch_loss=1.74]
> 100%|█████████████| 50/50 [00:12<00:00,  3.88it/s, epoch=5/5, avg_epoch_loss=1.58]
> ```
> 
> If you really need to train for `prediction_length=365`, I advise you to use longer training data (at least 2-3 years).

Thanks for your reply @lostella . I think I get the point in your reply, I will try to use longer data for training. And I still have one more question: why predict_length has such influence, in my opinion, I think the `context_length` may have this kind of influence, which I mean the model uses `context_length` data to train and predict  `prediction_length`  times, am I get the right point here? Cuz deepar is a one-step predict method. 
And thanks again.

**Comment by lostella** [comment]
In its heart it is a one-step prediction method (based on a RNN), but it's trained by unrolling it for both context **and** prediction length, and computing the loss over the whole length. 

So with the data in your original example, the only data the model can see at training time is a slice that has the "prediction" segment all over the time series, and the "context" segment before that (with no data, all padding). For this reason, there essentially no chance to get a good 1-year prediction model with 1-year data only.

(This is in addition to any other consideration: with 1-year data the model essentially has only a single example of how a year goes, is that enough?)

**Comment by creamiracle** [comment]
> 

Thanks for explaination, I think I get the main point here that is `all the data use for context is padding 0`.  For the extra question, it seems not enough data, but in my situation, items are not selled at the same time, but they all need to be predicted for a same length (such as 365 days) so I use all items' data together to train data, then get a wierd loss. After checking each item's series, I find the situation before is the bad case, so I make this issue for that.
Do you think use different model for different data is a good idea to solve this situation? such as predict 30 days 12 times instead of predict 365 (a year) days at one time.
Thanks.

**Comment by lostella** [comment]
I think it would make sense to consider shorter prediction length in this case. If you want to get a prolonged prediction, you can train a model with a prediction length of 30 days; then predict for 30 days, feed the prediction back to the model (for example, the median prediction), and get a prediction for 30 more days, and so on.

However if you do that, the later predictions for the year should be expected to be less and less accurate, since any deviation from the ground truth will accumulate over time. If you have longer data though, you could do some back testing to check how well this strategy of feeding predictions back into the model work.
